sdk/iothub/src/service/requests/create_or_update_device_identity_builder.rs

```diff
@@ -0,0 +1,183 @@
+    pub(crate) fn new(
+        service_client: &'a ServiceClient,
+        operation: IdentityOperation,
+        etag: Option<String>,
+    ) -> Self {
+        Self {
+            service_client,
+            authentication: None,
+            capabilities: DeviceCapabilities::default(),
+            device_id: None,
+            status: None,
+            etag,
+            operation,
+        }
+    }
+
+    /// Set authentication to SAS on the device
+    pub fn authentication_using_sas<S, T>(mut self, primary_key: S, secondary_key: T) -> Self
+    where
+        S: Into<String>,
+        T: Into<String>,
+    {
+        self.authentication = Some(AuthenticationMechanism {
+            authentication_type: AuthenticationType::SAS,
+            x509_thumbprint: X509ThumbPrint::default(),
+            symmetric_key: SymmetricKey {
+                primary_key: Some(primary_key.into()),
+                secondary_key: Some(secondary_key.into()),
+            },
+        });
+
+        self
+    }
+
+    /// Set authentication to x509 on the device
+    pub fn authentication_using_x509<S, T>(
+        mut self,
+        primary_thumbprint: S,
+        secondary_thumbprint: T,
+    ) -> Self
+    where
+        S: Into<String>,
+        T: Into<String>,
+    {
+        self.authentication = Some(AuthenticationMechanism {
+            authentication_type: AuthenticationType::SelfSigned,
+            x509_thumbprint: X509ThumbPrint {
+                primary_thumbprint: Some(primary_thumbprint.into()),
+                secondary_thumbprint: Some(secondary_thumbprint.into()),
+            },
+            symmetric_key: SymmetricKey::default(),
+        });
+
+        self
+    }
+
+    /// Set authentication to certificate authority on the device
+    pub fn authentication_using_certificate_authority(mut self) -> Self {
+        self.authentication = Some(AuthenticationMechanism {
+            authentication_type: AuthenticationType::Authority,
+            x509_thumbprint: X509ThumbPrint::default(),
+            symmetric_key: SymmetricKey::default(),
+        });
+
+        self
+    }
+
+    /// Sets a device capability on the device
+    pub fn device_capability(mut self, desired_capability: DesiredCapability) -> Self {
+        match desired_capability {
+            DesiredCapability::IotEdge => self.capabilities.iotedge = true,
+        }
+        self
+    }
+
+    /// Sets the device id of the device
+    pub fn device_id<S>(mut self, device_id: S) -> Self
+    where
+        S: Into<String>,
+    {
+        self.device_id = Some(device_id.into());
+        self
+    }
+
+    /// Sets the desired status of the device
+    pub fn status(mut self, status: Status) -> Self {
+        self.status = Some(status);
+        self
+    }
+
+    /// Performs the create or update request on the device identity
+    pub async fn execute(self) -> Result<DeviceIdentityResponse, IoTHubError> {
+        let device_id = self.device_id.ok_or_else(|| {
```

Comment:
For these required fields, we've been typically taking them as [arguments to the execute function](https://github.com/Azure/azure-sdk-for-rust/blob/02949f433aeb2ac58ada7063851cb3921b82e4f9/sdk/cosmos/src/requests/create_collection_builder.rs#L48) when they are data related and [arguments to the Builder's `new` function](https://github.com/Azure/azure-sdk-for-rust/blob/02949f433aeb2ac58ada7063851cb3921b82e4f9/sdk/cosmos/src/requests/create_collection_builder.rs#L21) when they are configuration related.